File: tools/pplx-exporter/pplx_exporter/render.py
"""Render thread data to Markdown with YAML frontmatter."""

from __future__ import annotations

import hashlib
import re

from .config import BASE_URL
from .types import ThreadData


def render_thread(thread: ThreadData) -> tuple[str, str, int]:
    """Render a thread to Markdown with YAML frontmatter.

    Returns (markdown_string, sha256_hash, byte_count).
    """
    frontmatter = _build_frontmatter(thread)
    body = _build_body(thread)
    markdown = f"---\n{frontmatter}---\n\n{body}"

    encoded = markdown.encode("utf-8")
    sha = hashlib.sha256(encoded).hexdigest()
    return markdown, f"sha256:{sha}", len(encoded)


def _build_frontmatter(thread: ThreadData) -> str:
    """Build YAML frontmatter block."""
    lines: list[str] = []
    lines.append(f"pplx_slug: {thread.slug}")
    lines.append(f"pplx_url: {BASE_URL}/search/{thread.slug}")
    lines.append(f"title: {_yaml_quote(thread.title)}")

    if thread.space:
        lines.append(f"space: {_yaml_quote(thread.space)}")

    if thread.created_at:
        lines.append(f"created_at: \"{thread.created_at}\"")
    if thread.updated_at:
        lines.append(f"updated_at: \"{thread.updated_at}\"")

    lines.append("source: perplexity")
    lines.append("tier: INTUITED")
    lines.append("export_tool: pplx-exporter-v0.1")
    lines.append("")

    return "\n".join(lines)


def _build_body(thread: ThreadData) -> str:
    """Build the conversation body."""
    parts: list[str] = []
    parts.append(f"# {thread.title}\n")

    for i, entry in enumerate(thread.entries, 1):
        if entry.query:
            parts.append(f"## Q{i}: {entry.query}\n")
        if entry.answer:
            parts.append(entry.answer)
            parts.append("")

        if entry.sources:
            parts.append("### Sources\n")
            for src in entry.sources:
                title = src.get("title", "")
                url = src.get("url", "")
                if url:
                    parts.append(f"- [{title}]({url})")
                elif title:
                    parts.append(f"- {title}")
            parts.append("")

    return "\n".join(parts)


def _yaml_quote(value: str) -> str:
    """Quote a string for YAML if it contains special characters."""
    if any(c in value for c in (':', '#', '"', "'", '{', '}', '[', ']', ',', '&', '*', '?', '|', '>', '!', '%', '@')):
        escaped = value.replace('"', '\\"')
        return f'"{escaped}"'
    if not value or value != value.strip():
        return f'"{value}"'
    return f'"{value}"'


def slugify_filename(title: str) -> str:
    """Convert a title to a filesystem-safe filename."""
    clean = re.sub(r'[<>:"/\\|?*]', "", title)
    clean = re.sub(r"\s+", "-", clean.strip())
    clean = clean[:120]
    return clean or "untitled"

File: tools/pplx-exporter/src/render.ts
import { createHash } from "node:crypto";
import type { ConversationResponse, Entry, Block, SourceRow } from "./types.js";

interface RenderOptions {
  slug: string;
  spaceName?: string;
  spaceUuid?: string;
}

export interface RenderedThread {
  markdown: string;
  sha256: string;
  byteCount: number;
  title: string;
  createdAt?: string;
  updatedAt?: string;
}

export function renderThread(conversation: ConversationResponse, opts: RenderOptions): RenderedThread {
  const title = conversation.thread_title ?? conversation.entries?.[0]?.thread_title ?? "Untitled";
  const updatedAt = conversation.updated_datetime ?? conversation.entries?.[0]?.updated_datetime;
  const createdAt = findEarliestDate(conversation.entries);
  const spaceName = opts.spaceName ?? conversation.collection_info?.title;
  const spaceUuid = opts.spaceUuid ?? conversation.collection_info?.uuid;

  const frontmatter = buildFrontmatter({
    pplx_slug: opts.slug,
    pplx_url: `https://www.perplexity.ai/search/${opts.slug}`,
    title,
    space: spaceName,
    space_uuid: spaceUuid,
    created_at: createdAt,
    updated_at: updatedAt,
    source: "perplexity",
    tier: "INTUITED",
    tier_note: "Exported from Perplexity UI. Content not independently verified.",
    export_tool: "pplx-exporter-v0.1",
  });

  const body = renderEntries(conversation.entries);
  const markdown = frontmatter + "\n" + body;
  const byteCount = Buffer.byteLength(markdown, "utf-8");
  const sha256 = createHash("sha256").update(markdown).digest("hex");

  return { markdown, sha256, byteCount, title, createdAt, updatedAt };
}

function buildFrontmatter(fields: Record<string, string | undefined>): string {
  const lines = ["---"];
  for (const [key, value] of Object.entries(fields)) {
    if (value === undefined || value === null) continue;
    const escaped = value.includes(":") || value.includes('"') ? `"${value.replace(/"/g, '\\"')}"` : value;
    lines.push(`${key}: ${escaped}`);
  }
  lines.push("---");
  return lines.join("\n");
}

function renderEntries(entries: Entry[]): string {
  if (!entries || entries.length === 0) return "*No conversation content found.*\n";

  const sections: string[] = [];

  for (const entry of entries) {
    if (entry.query_str) {
      sections.push(`## ${entry.query_str}\n`);
    }

    if (entry.blocks && Array.isArray(entry.blocks)) {
      for (const block of entry.blocks) {
        sections.push(renderBlock(block));
      }
    }
  }

  return sections.join("\n");
}

function renderBlock(block: Block): string {
  switch (block.intended_usage) {
    case "ask_text":
      return (block.markdown_block?.answer ?? "") + "\n";

    case "sources_answer_mode":
      return renderSources(block.sources_mode_block?.rows ?? []);

    case "image_answer_mode":
      return renderImages(block.image_mode_block?.media_items ?? []);

    case "video_answer_mode":
      return renderVideos(block.video_mode_block?.media_items ?? []);

    default:
      if (block.markdown_block?.answer) {
        return block.markdown_block.answer + "\n";
      }
      return "";
  }
}

function renderSources(sources: SourceRow[]): string {
  if (sources.length === 0) return "";

  const lines = ["\n### Sources\n"];
  for (let i = 0; i < sources.length; i++) {
    const s = sources[i];
    const title = s.title ?? s.url ?? `Source ${i + 1}`;
    const url = s.url ?? "#";
    lines.push(`${i + 1}. [${title}](${url})`);
  }
  return lines.join("\n") + "\n";
}

function renderImages(items: { url?: string; alt?: string }[]): string {
  if (items.length === 0) return "";

  const lines = items.map((img) => `![${img.alt ?? "image"}](${img.url ?? ""})`);
  return lines.join("\n") + "\n";
}

function renderVideos(items: { url?: string; title?: string }[]): string {
  if (items.length === 0) return "";

  const lines = items.map((v) => `[${v.title ?? "Video"}](${v.url ?? ""})`);
  return lines.join("\n") + "\n";
}

function findEarliestDate(entries: Entry[]): string | undefined {
  if (!entries || entries.length === 0) return undefined;

  const dates = entries
    .map((e) => e.updated_datetime)
    .filter(Boolean)
    .sort();

  return dates[0] ?? undefined;
}
